### ISSUE TITLE
Fiji crashes on startup after update with some kind of "Access Denied" error

### ISSUE THREAD

**Issue by haesleinhuepf** [fiji/fiji#257]
Hi all,

as discussed with @ndefrancesco on [image.sc](https://forum.image.sc/t/console-opens-when-starting-fiji-on-windows/41093/3) Fiji starts now with a black console window showing an error message. Some installations also freeze at that point:
![image](https://user-images.githubusercontent.com/12660498/89126911-f4568200-d4e9-11ea-80f8-8d91523bd07c.png)

The error number "5" often stands for "Access denied" and Nicos suggestion of running Fiji with administrator priviledges would solve this issue confirm that. However, in the light of recent events to the server where the update sites are hosted, I would prefer running Fiji without administrator preiviledges.

1) Does anyone know what Fiji is trying to change at the System configuration?

2) Can we undo the rollout of this exe file and stick to the old one until the issue is solved?

Thanks!

Cheers,
Robert

**Comment by haesleinhuepf** [comment]
While the win64-exe issue is resolved for the moment (I uploaded the old exe to the update site as @ctrueden suggested in the forum), I'm having issues with running the 32-bit version. When updating the [32-Bit Fiji from July 11th](https://downloads.imagej.net/fiji/archive/20200711-0516/fiji-win32.zip) it doesn't start anymore. This happens before updating the exe file. While ImageJ-win32.exe (again, that's the 20200711 version) doesn't tell any error, renaming it to debug.exe reveals this:

![image](https://user-images.githubusercontent.com/12660498/89163550-0aaa1f80-d576-11ea-8bde-b20d0c9506f9.png)

If anybody has an idea what to do next, hints are welcome. I can dive deeper into it by the end of the week. I hope there are not too many win32-users suffering from this.

Cheers,
Robert

**Comment by ctrueden** [comment]
@haesleinhuepf wrote:
> Could not reserve enough space for 1253376KB object heap

That's the dreaded "trying to allocate more memory than a 32-bit JVM supports on Windows" bug. You'll notice it's only 1.2GB: seemingly far less than the 2GB limit, right? But the 32-bit Windows JVM is notorious for actually allowing much less as a maximum; IIRC memory is shared with Windows itself? So the cap varies depending on your system and version of Windows. No one (to my knowledge) ever found a way to safely detect this upper bound on a per-system basis. But historically, 1.2GB was pretty safe, so the launcher caps it there—but I guess it's not good enough anymore on modern systems. Try lowering it to 1.1GB or even 1.0GB? You should be able to hack your ImageJ.cfg to change the limit.

Of course, hacking your ImageJ.cfg locally will not fix the launcher for everyone in a similar situation. But in order to do that, we'd need to publish a new version of the 32-bit launcher with a lower limit.

I'm guessing this has been an issue for quite some time for people who continue using 32-bit systems. The fact that no one has reported it before is a good sign, suggesting that either: A) it only affects a small fraction of win32 systems; or B) there are very few people using win32 in general these days.

**Comment by haesleinhuepf** [comment]
> I'm guessing this has been an issue for quite some time for people who continue using 32-bit systems. The fact that no one has reported it before is a good sign, suggesting that either: A) it only affects a small fraction of win32 systems; or B) there are very few people using win32 in general these days.

I agree. Among my students in the last course, 1out of 60 had a Windows 32-bit running. She was happy to finally have a good reason to get a new laptop. Is it maybe time to discontinue the 32-bit version? I would vote for it and defend it in case of trouble. I also recently argued why CLIJ may not work on Windows 7 :-)

**Comment by ctrueden** [comment]
I would be happy to retire the win32 support. We already retired linux32. (I don't remember exactly why, but maybe due to difficulties building a linux32 launcher easily on Travis CI? Or maybe just because no one uses linux32 anymore?)

Do you see any issue with continuing to ship the last working ImageJ-win32.exe? This is what we do for ImageJ-linux32. It's a "soft" discontinuation, in that people are free to continue using those launchers if they work, but we won't be updating them anymore.

**Comment by haesleinhuepf** [comment]
> Do you see any issue with continuing to ship the last working ImageJ-win32.exe? This is what we do for ImageJ-linux32.

No issue and it sounds like a great idea. I just can't upload it right now, because Fiji doesn't start with this launcher... Can one upload a windows imagej-win32.exe from a linux Fiji?

**Comment by ctrueden** [comment]
> Can one upload a windows imagej-win32.exe from a linux Fiji?

Yep, you can upload any platform-specific thing from any running Fiji, even if the platform is different.

In the details box, just make sure it still says the platform is win32. But since the file already exists and is flagged that way, I think it should already be set.

If you have trouble let me know; I can also just revert it directly in the `db.xml.gz`.

**Comment by ctrueden** [comment]
@haesleinhuepf I discussed with @hinerm. Here is how we want to proceed with the two issues discussed here.

## ImageJ-win64

After some analysis, here is what @hinerm and I have concluded:

1. The new launcher has a regression where the console always pops, even when not passing the `--console` flag. This needs to be fixed, but it's not the reason for Fiji hanging at startup.

2. The `Could not open/create prefs root node Software\JavaSoft\Prefs at root 0x80000002. Windows RegCreateKeyEx(...) returned error code 5.` message is just a warning, and not the reason for Fiji hanging at startup. It's a [bug in Java 8](https://bugs.openjdk.java.net/browse/JDK-8139507), fixed in Java 1.8.0_202 (see also [this SO answer](https://stackoverflow.com/a/41151051/1207769)). This error message is _not_ a new bug in the launcher—the old launchers also display this message. It's been [mentioned here on the forum several times before](https://forum.image.sc/search?q=could%20not%20open%2Fcreate%20prefs). The reason people weren't (in general) noticing the message before, but are seeing it now, is due to (1) above: previously, the console was not popping up without the `--console` flag being given.

3. The fact that Fiji hangs at startup on some systems with the new launcher, but not with the old launcher, is a critical new bug, which we still don't understand. @haesleinhuepf On your affected system, please rename `ImageJ-win64.exe` to `debug.exe` and run it, to see what's happening in more detail. If there's nothing helpful there, you can then generate a thread dump (<kbd>ctrl</kbd>+</kbd>pause</kbd>) to see if things are deadlocked on the Java side. Please paste any interesting findings here on this issue.

## ImageJ-win32

I pushed a fix (imagej/imagej-launcher@96e38eac270bb1e146f63a65c535880ea562087c) for the `Could not reserve enough space` error: I dropped the default max heap for 32-bit systems from 1224 to 1024 MB. Please download the [latest SNAPSHOT of ImageJ-win32.exe](https://maven.scijava.org/service/local/artifact/maven/redirect?r=snapshots&g=net.imagej&a=imagej-launcher&v=5.0.4-SNAPSHOT&e=exe&c=win32) and give it a try on the affected system.

## TL;DR

* For 64-bit, please run the [new ImageJ-win64.exe](https://maven.scijava.org/service/local/artifact/maven/redirect?r=releases&g=net.imagej&a=imagej-launcher&v=5.0.3&e=exe&c=win64) with `debug.exe` and generate a thread dump if needed.
* For 32-bit, please try the [latest `ImageJ-win32.exe` SNAPSHOT](https://maven.scijava.org/service/local/artifact/maven/redirect?r=snapshots&g=net.imagej&a=imagej-launcher&v=5.0.4-SNAPSHOT&e=exe&c=win32).

**Comment by haesleinhuepf** [comment]
Alright. I just download and tested them. The new win64.exe works on the system. It's just not signed properly I guess:
![image](https://user-images.githubusercontent.com/12660498/89327630-7e3e5080-d68c-11ea-92fe-ba8da0a73282.png)

As I trust you, I ran it anyway and it produced this output:
https://gist.github.com/haesleinhuepf/f93cef9e4630180dda26aa1b1dfd3690

Then, I ran its update (it downloads updater.jar) and restarted it. While restarting, it freezes again. 
![image](https://user-images.githubusercontent.com/12660498/89327990-04f32d80-d68d-11ea-9666-c9d517881a5d.png)

The menu is reactive but when I click any menu, nothing happens. Here is its output, I'm afraid the first part is missing:
https://gist.github.com/haesleinhuepf/54598799831e7bb2e4762a686f8c1c71

After restarting, it works again. I updated to download also all the jars. Here its output.
https://gist.github.com/haesleinhuepf/b61ef25fef5272913f06a9040a5669db

After this second update also the finally updated Fiji works. Here comes its output:
https://gist.github.com/haesleinhuepf/a58e7065f02aa7ccca0c720156dbd215


Last but not least I put the win32-exe in a win-32 installation (I think I didn't do it like that last time):
https://gist.github.com/haesleinhuepf/0dd4f3839001805a06bdc81de01aef15

It also runs fine. I updated it and it's still ok.

Looks like great work!

Let me know if I can help with anything else!

Cheers,
Robert

**Comment by haesleinhuepf** [comment]
I also tried with some more older Fijis and they appear to work fine with the new launcher. 👍

**Comment by ctrueden** [comment]
Thanks a lot @haesleinhuepf for testing. So... what can we do to help improve the experience for users? I don't know what to do about this Windows Defender SmartScreen thing. And why didn't you see the "Windows protected this PC" last time you tested it? We didn't change anything yet...

**Comment by haesleinhuepf** [comment]
> I don't know what to do about this Windows Defender SmartScreen thing.

It's a bit weird. I only saw it once.

> we can release 5.0.4 and test it, and hopefully it works and we upload it?

I'm again happy to test it.

**Comment by refack** [comment]
> 3\. The fact that Fiji hangs at startup on some systems with the new launcher, but not with the old launcher, is a critical new bug, which we still don't understand. @haesleinhuepf On your affected system, please rename `ImageJ-win64.exe` to `debug.exe` and run it, to see what's happening in more detail. If there's nothing helpful there, you can then generate a thread dump (ctrl+pause) to see if things are deadlocked on the Java side. Please paste any interesting findings here on this issue.

I'm seeing the above mentioned quick crash issue after updating just now (Windows 10 64bit Insider build 20185).
Following is the console output after renaming to `debug.exe`:
```console
Detected Fiji
Using JRE from ImageJ.cfg: D:\bin\tools\fiji/jre
Found Java options in ImageJ.cfg: '-Xmx7560m -cp ij.jar'
JRE not found in 'D:\bin\tools\fiji/jre'
JAVA_HOME contains a JRE: 'D:\bin\tools\fiji/java/win64/jdk1.8.0_172//jre'
java -Xmx7560m -Dpython.cachedir.skip=true -Dplugins.dir=D:\\bin\\tools\\fiji -Xincgc -XX:PermSize=128m -Dsun.java.command=ImageJ -Djava.class.path=D:\\bin\\tools\\fiji/jars/imagej-launcher-5.0.3.jar -Dimagej.dir=D:\\bin\\tools\\fiji -Dij.dir=D:\\bin\\tools\\fiji -Dfiji.dir=D:\\bin\\tools\\fiji -Dfiji.defaultLibPath=bin/server/jvm.dll -Dfiji.executable=D:\\bin\\tools\\fiji\\debug.exe -Dij.executable=D:\\bin\\tools\\fiji\\debug.exe -Djava.library.path=D:\\bin\\tools\\fiji/lib/win64;D:\\bin\\tools\\fiji/mm/win64 -Dsun.java2d.noddraw=true -Dij.debug=true -Dscijava.log.level=debug net.imagej.launcher.ClassLauncher -classpath D:\\bin\\tools\\fiji/jars/ij-1.53c.jar fiji.Main -port7
Using JAVA_HOME D:\bin\tools\fiji/java/win64/jdk1.8.0_172//jre
Opening Java library D:\bin\tools\fiji/java/win64/jdk1.8.0_172//jre/bin/server/jvm.dll
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option PermSize=128m; support was removed in 8.0
Java HotSpot(TM) 64-Bit Server VM warning: Using incremental CMS is deprecated and will likely be removed in a future release
Adding option: -Djava.home=D:\bin\tools\fiji/java/win64/jdk1.8.0_172/
Launching main class fiji.Main with parameters [-port7]
Class loader = sun.misc.Launcher$AppClassLoader@c387f44
java.lang.ClassNotFoundException: fiji.Main
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at net.imagej.launcher.ClassLauncher.launch(ClassLauncher.java:264)
        at net.imagej.launcher.ClassLauncher.run(ClassLauncher.java:198)
        at net.imagej.launcher.ClassLauncher.main(ClassLauncher.java:89)
java.lang.NullPointerException
        at net.imagej.launcher.ClassLauncher.launch(ClassLauncher.java:281)
        at net.imagej.launcher.ClassLauncher.run(ClassLauncher.java:198)
        at net.imagej.launcher.ClassLauncher.main(ClassLauncher.java:89)
```
